alpha/apps/kaltura/lib/microservice/MicroServiceAppRegistry.php

```diff
@@ -4,13 +4,156 @@
  */
 class MicroServiceAppRegistry extends MicroServiceBaseService
 {
+	const APP_GUID_CACHE_KEY_PREFIX = 'appGuid_';
+	const APP_GUID_CACHE_TTL = 86400; // cached for 1 day
+	
 	public function __construct()
 	{
 		parent::__construct('app-registry','app-registry');
 	}
+	
+	/**
+	 * @throws kCoreException
+	 */
+	public function get($partnerId, $appGuid)
+	{
+		return $this->serve($partnerId,'get', array('id' => $appGuid));
+	}
 
 	public function list($partnerId, $filter, $pager = array())
 	{
 		return $this->serve($partnerId,'list', array('filter' => $filter, 'pager' => $pager));
 	}
-}
\ No newline at end of file
+	
+	/**
+	 * @throws kCoreException
+	 */
+	public static function getExistingAppGuid($partnerId, $appGuid)
+	{
+		$appGuidExists = MicroServiceAppRegistry::getAppGuidFromCache($appGuid);
+		if ($appGuidExists)
+		{
+			return $appGuidExists;
+		}
+		
+		$appRegistryClient = new MicroServiceAppRegistry();
+		$appRegistry = $appRegistryClient->get($partnerId, $appGuid);
+		
+		if (isset($appRegistry->code) && $appRegistry->code == 'OBJECT_NOT_FOUND')
+		{
+			return false;
+		}
+		
+		if (!isset($appRegistry->id) || !kString::isValidMongoId($appRegistry->id))
+		{
+			return false;
+		}
+		
+		MicroServiceAppRegistry::addAppGuidToCache($appRegistry->id);
+		return $appRegistry->id;
+	}
+	
+	/**
+	 * @throws kCoreException
+	 */
+	public static function getExistingAppGuidsFromCsv($partnerId, $appGuidsCsv)
+	{
+		$appGuidsResponse = array();
+		$appGuidsCached = array();
+		$appGuidsToQuery = explode(',', $appGuidsCsv);
+		
+		// remove duplicate appGuids, if exists
+		$appGuidsToQuery = array_unique($appGuidsToQuery);
+		
+		// filter appGuids that already stored in cache
+		foreach ($appGuidsToQuery as $key => $appGuid)
+		{
+			$appGuidExist = MicroServiceAppRegistry::getAppGuidFromCache($appGuid);
+			
+			// remove appGuids that already stored in cache before querying app-registry ms
+			if ($appGuidExist)
+			{
+				$appGuidsCached[] = $appGuid;
+				unset($appGuidsToQuery[$key]);
+			}
+		}
+		
+		// reset array keys or else 'json_encode' will encode the non-consecutive as the key-value resulting in bad mongo query
+		$appGuidsToQuery = array_values($appGuidsToQuery);
+		
+		// if all appGuids were found in cache, return original csv
+		if (!count($appGuidsToQuery))
+		{
+			return $appGuidsCsv;
+		}
+		
+		$filter = array(
+			'idIn' => $appGuidsToQuery
+		);
+		
+		$pager = array(
```

Comment:
Did you check if we allow large APIs if, for example, we get 10000 ids?
And if so how the MS will behave in such a case?

Reply:
Generally speaking, EP will send a single appGuid most times (per EP site since appGuid represent an EP instance)
I do not believe that the appGuid is going to be the major bottleneck

Reply:
@yossipapi spoke with Daniel S. the dev working on it
He confirmed the use case is going to be 2 appGuids most time

This we should do something else here?
I can open a separate ticket to do a small 'load' test of sending 1000 appGuids

Not sure we will ever get 10000